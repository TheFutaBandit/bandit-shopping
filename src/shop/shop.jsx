import { useState, useEffect } from "react";
import PropTypes from 'prop-types';
import cart from '../assets/cart.svg';
import negative from '../assets/negative.svg'
import positive from '../assets/positive.svg'
import { Cart_styled, Card_Container_Styled, FooterText, Footer_Styled, Card_Styled, ContentText } from './shop.styled.js'

export const Footer = () => {
    return (
        <Footer_Styled>
            <FooterText size = "1rem" weight = "900">WE HAVE ALL THE RIGHTS YOU KNOW?</FooterText>
            <FooterText size = "0.8rem" weight = "400">DONâ€™T TRY ANYTHING FUNNY.</FooterText>
        </Footer_Styled>
    )
}

export const Cart = ({counter}) => {
    return (
        <Cart_styled>
            <img src = {cart} />
            <p className = "buy_text">Click To Steal!</p>
            <div data-testid="cart_counter" className = "cart_counter">{counter}</div>
        </Cart_styled>
    )
}

export const Card = ({
        incrementCounter, 
        decrementCounter,
        prodLink = "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg", 
        prodTitle = "Mens Casual \nSlim Fit", 
        prodCost = 500, 
        prodQuantity = 0,
    }) => {
    const[quantity, setQuantity] = useState(0);
    let fontSize = "1.1rem";

    return (
        <Card_Styled>
            <div className = "imageContainer">
                <img src = {prodLink}></img>
            </div>
            <div className = "contentContainer">
                <div className = "content">
                    <ContentText size = {fontSize} weight = "300">{prodTitle}</ContentText>
                    <ContentText size = "1.1rem" weight = "300">${prodCost}</ContentText>
                </div>
            </div>
            <div className = "counter">
                <img alt="decrement_button" src = {negative} onClick={() => {
                    if(quantity > 0) {
                        decrementCounter()
                        setQuantity((prev) => prev-1);
                    }
                }}></img>
                <FooterText data-testid = "card_counter" size = "1.5rem" weight = "100">{quantity}</FooterText>
                <img alt="increment_button" src = {positive} onClick={() => {
                    incrementCounter()
                    setQuantity((prev) => prev+1);
                }}></img>
            </div>
        </Card_Styled>
    )
}

Card.propTypes = {
    incrementCounter: PropTypes.func,
    decrementCounter: PropTypes.func,
    prodLink: PropTypes.string,
    prodTitle: PropTypes.string,
    prodCost: PropTypes.number,
    prodQuantity: PropTypes.number
}

export const Card_Container = ({incrementCounter, decrementCounter}) => {
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(true);
    const [storeUrl, setStoreUrl] = useState('https://fakestoreapi.com/products');
    const [productList, setProductList] = useState([]);

    useEffect(() => {
        const getData = async () => {
            try {
                const response = await fetch(storeUrl, { mode: "cors" });
                if(response.status >= 400) {
                    throw new Error("server error");
                }
                const data = await response.json();
                console.log(data);
                setProductList(JSON.parse(JSON.stringify(data.slice(0,4))));
                setLoading(false);
                
            } catch (error) {
                setError(error);
                setLoading(false);
            }
        };
        getData();   
        
    }, [storeUrl]);

    if(loading) return <p>Loading...</p>;
    if(error) return <p>A network error was encountered</p>;
    console.log(productList)

    return (
        <Card_Container_Styled>
            {productList.map((item) => {
                return (
                    <Card 
                        key = {item.id}
                        incrementCounter = {incrementCounter} 
                        decrementCounter = {decrementCounter} 
                        prodLink = {item.image}
                        prodCost = {item.price}
                        prodTitle = {item.title}
                    />
                )
            })}
        </Card_Container_Styled>
    )
}

export const Shop = () => {
    const [counter, setCounter] = useState(0);

    function incrementCounter() {
        setCounter((prev) => prev+1);
    }

    function decrementCounter() {
        setCounter((prev) => prev-1);
    }

    return (
        <>
            <Cart counter={counter}/>
            <Card_Container incrementCounter = {incrementCounter} decrementCounter = {decrementCounter}/>
            <Footer />
        </>
    )
}

