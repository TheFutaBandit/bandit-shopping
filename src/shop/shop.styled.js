import {styled} from 'styled-components'

export const Cart_styled = styled.div`
    display: flex;
    justify-content: flex-start;
    position: relative;
    margin-top: 2rem;
    cursor: pointer;

    img {
        min-width: 4rem;
        height: auto;
    }

    .buy_text {
        position: absolute;
        top: -20px;
        left: -30px;
        transform: rotate(-45deg);
        font-size: 0.8rem;
        font-weight: 300;
    }

    .cart_counter {
        position: absolute;
        font-weight: 100;
        font-size: 1rem;
        top: 0;
        right: 0;
        border: 0.5px solid black;
        width: 22px;
        height: 22px;
        border-radius: 100%;
        background-color: white;
        display: flex;
        justify-content: center;
        align-items: center;
    }
`

export const Card_Container_Styled = styled.div`
    display: grid;
    max-width: 100%;
    grid-template-columns: repeat(auto-fit, 224px);
    margin-top: 2rem;
    column-gap: 32px;
    row-gap: 32px;
    margin-bottom: 2rem;
`
export const Card_Styled = styled.div`
    border: 1px solid black;
    height: 280px;
    display: flex;
    flex-direction: column;
    max-height: inherit;
    .imageContainer {
        flex: 0 0 184px;
        border-bottom: 1px solid black;
        display: flex;
        align-items: center;
        justify-content: center;
        img {
            max-width:100px;
            height: auto;
        }
    }
    .contentContainer {
        padding: 7px;
        .content {
            display: flex;
            justify-content: space-between;
            margin-bottom: 0px;
        }
        flex-grow: 1;
        overflow-y: auto;
    }
    .counter {
        display: flex;
        justify-self: flex-end;
        align-items: center;
        justify-content: center;
        gap: 8px;
        img {
            height: auto;
            max-width: 27px;
            cursor: pointer;
        }
        margin-bottom: 4px;
    }
    
`

export const FooterText = styled.p`
    font-size: ${prop => prop.size};
    font-weight: ${prop => prop.weight};
    margin: 0px;
`

export const ContentText = styled.p`
    font-size: ${prop => prop.size};
    font-weight: ${prop => prop.weight};
    max-width: 100px;
    white-space: pre-line;
    margin: 0px;
`

export const Footer_Styled = styled.div`
    display: flex;
    flex-direction: column;
    flex: 1;
    justify-content: flex-end;
    align-items: center;
    margin-bottom: 30px;
`


