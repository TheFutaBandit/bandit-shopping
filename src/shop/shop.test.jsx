import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {Shop} from './shop'

describe("Shop Render and Card", () => {
    it("Renders Shop Page", () => {
        const { container } = render(<Shop />);
        expect(container).toMatchSnapshot();
    });

    it("card quantity does not decrement below 0 after negative button clicked", async () => {
        const user = userEvent.setup();
        render(<Shop />)

        const button = screen.getAllByAltText("decrement_button");

        await user.click(button[0]);

        expect(screen.getAllByTestId("card_counter")[0].textContent).toMatch("0");
    })

    it("card quantity decrements after negative button clicked given its above 0", async () => {
        const user = userEvent.setup();
        render(<Shop />)

        const button = screen.getAllByAltText("decrement_button");
        const incrementButton = screen.getAllByAltText("increment_button");

        await user.click(incrementButton[0])

        await user.click(incrementButton[0])

        await user.click(button[0]);

        expect(screen.getAllByTestId("card_counter")[0].textContent).toMatch("1");
    })

    it("card quantity increases after positive button clicked", async () => {
        const user = userEvent.setup();
        render(<Shop />);

        const button = screen.getAllByAltText("increment_button");

        await user.click(button[0]);

        expect(screen.getAllByTestId("card_counter")[0].textContent).toMatch("1");
    })
});

describe("Testing Cart Increase/Decrease Functionality", () => {
    it("Card Increment increments the cart counter", async () => {
        const user = userEvent.setup();

        render(<Shop />);

        const button = screen.getAllByAltText("increment_button");

        await user.click(button[0]);

        expect(screen.getByTestId("cart_counter").textContent).toMatch("1");
    })

    it("Card Decrement does decrement cart below 0 by default", async () => {
        const user = userEvent.setup();

        render(<Shop />);

        const button = screen.getAllByAltText("decrement_button");

        await user.click(button[0]);

        expect(screen.getByTestId("cart_counter").textContent).toMatch("0");
    });

    it("Card Decrements cart given 2 items in card", async () => {
        const user = userEvent.setup();

        render(<Shop />);

        const incrementButton = screen.getAllByAltText("increment_button");
        const decrementButton = screen.getAllByAltText("decrement_button");

        await user.click(incrementButton[0]);
        await user.click(incrementButton[0]);

        await user.click(decrementButton[0]);

        expect(screen.getByTestId("cart_counter").textContent).toMatch("1");
    })
})