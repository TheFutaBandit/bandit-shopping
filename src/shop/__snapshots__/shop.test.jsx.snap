// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop Render and Card > Renders Shop Page 1`] = `
<div>
  <div
    class="sc-FEMpB eCjMBC"
  >
    <img
      src="data:image/svg+xml,%3csvg%20width='62'%20height='62'%20viewBox='0%200%2062%2062'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M15.5%205.16666L7.75%2015.5V51.6667C7.75%2053.0369%208.29434%2054.3511%209.26328%2055.32C10.2322%2056.289%2011.5464%2056.8333%2012.9167%2056.8333H49.0833C50.4536%2056.8333%2051.7678%2056.289%2052.7367%2055.32C53.7057%2054.3511%2054.25%2053.0369%2054.25%2051.6667V15.5L46.5%205.16666H15.5Z'%20stroke='black'%20stroke-width='0.416667'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20d='M7.75%2015.5H54.25'%20stroke='black'%20stroke-width='0.416667'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3cpath%20d='M41.3334%2025.8333C41.3334%2028.5739%2040.2447%2031.2022%2038.3069%2033.1401C36.369%2035.078%2033.7407%2036.1667%2031.0001%2036.1667C28.2595%2036.1667%2025.6312%2035.078%2023.6933%2033.1401C21.7554%2031.2022%2020.6667%2028.5739%2020.6667%2025.8333'%20stroke='black'%20stroke-width='0.416667'%20stroke-linecap='round'%20stroke-linejoin='round'/%3e%3c/svg%3e"
    />
    <div
      class="cart_counter"
      data-testid="cart_counter"
    >
      0
    </div>
  </div>
  <div
    class="sc-gjcoXW jPFSgd"
  >
    <div
      class="sc-dVBluf bMZbZD"
    >
      <div
        class="imageContainer"
      >
        <img
          src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
        />
      </div>
      <div
        class="contentContainer"
      >
        <div
          class="content"
        >
          <p
            class="sc-fLDLck iiVBBO"
            weight="300"
          >
            Mens Casual 
Slim Fit
          </p>
          <p
            class="sc-fLDLck iiVBBO"
            weight="300"
          >
            $
            500
          </p>
        </div>
      </div>
      <div
        class="counter"
      >
        <img
          alt="decrement_button"
          src="data:image/svg+xml,%3csvg%20width='30'%20height='30'%20viewBox='0%200%2030%2030'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3ccircle%20cx='15'%20cy='15'%20r='14.25'%20fill='white'%20stroke='black'%20stroke-width='0.5'/%3e%3cline%20x1='10.25'%20y1='15.75'%20x2='19.75'%20y2='15.75'%20stroke='black'%20stroke-width='0.5'%20stroke-linecap='round'/%3e%3c/svg%3e"
        />
        <p
          class="sc-fLDLck PSnoQ"
          data-testid="card_counter"
          weight="100"
        >
          0
        </p>
        <img
          alt="increment_button"
          src="data:image/svg+xml,%3csvg%20width='30'%20height='30'%20viewBox='0%200%2030%2030'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3ccircle%20cx='15'%20cy='15'%20r='14.25'%20fill='white'%20stroke='black'%20stroke-width='0.5'/%3e%3cline%20x1='10.25'%20y1='14.75'%20x2='19.75'%20y2='14.75'%20stroke='black'%20stroke-width='0.5'%20stroke-linecap='round'/%3e%3cline%20x1='15.25'%20y1='10.25'%20x2='15.25'%20y2='19.75'%20stroke='black'%20stroke-width='0.5'%20stroke-linecap='round'/%3e%3c/svg%3e"
        />
      </div>
    </div>
    <div
      class="sc-dVBluf bMZbZD"
    >
      <div
        class="imageContainer"
      >
        <img
          src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
        />
      </div>
      <div
        class="contentContainer"
      >
        <div
          class="content"
        >
          <p
            class="sc-fLDLck iiVBBO"
            weight="300"
          >
            Mens Casual 
Slim Fit
          </p>
          <p
            class="sc-fLDLck iiVBBO"
            weight="300"
          >
            $
            500
          </p>
        </div>
      </div>
      <div
        class="counter"
      >
        <img
          alt="decrement_button"
          src="data:image/svg+xml,%3csvg%20width='30'%20height='30'%20viewBox='0%200%2030%2030'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3ccircle%20cx='15'%20cy='15'%20r='14.25'%20fill='white'%20stroke='black'%20stroke-width='0.5'/%3e%3cline%20x1='10.25'%20y1='15.75'%20x2='19.75'%20y2='15.75'%20stroke='black'%20stroke-width='0.5'%20stroke-linecap='round'/%3e%3c/svg%3e"
        />
        <p
          class="sc-fLDLck PSnoQ"
          data-testid="card_counter"
          weight="100"
        >
          0
        </p>
        <img
          alt="increment_button"
          src="data:image/svg+xml,%3csvg%20width='30'%20height='30'%20viewBox='0%200%2030%2030'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3ccircle%20cx='15'%20cy='15'%20r='14.25'%20fill='white'%20stroke='black'%20stroke-width='0.5'/%3e%3cline%20x1='10.25'%20y1='14.75'%20x2='19.75'%20y2='14.75'%20stroke='black'%20stroke-width='0.5'%20stroke-linecap='round'/%3e%3cline%20x1='15.25'%20y1='10.25'%20x2='15.25'%20y2='19.75'%20stroke='black'%20stroke-width='0.5'%20stroke-linecap='round'/%3e%3c/svg%3e"
        />
      </div>
    </div>
  </div>
  <div
    class="sc-dycHyt knBunj"
  >
    <p
      class="sc-fLDLck emOOIz"
      weight="900"
    >
      WE HAVE ALL THE RIGHTS YOU KNOW?
    </p>
    <p
      class="sc-fLDLck gnPpXx"
      weight="400"
    >
      DONâ€™T TRY ANYTHING FUNNY.
    </p>
  </div>
</div>
`;
